TicTacToe/TicTacToe/Controllers/GamePageViewController.swift

```diff
@@ -0,0 +1,116 @@
+            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
+                self.aiTurn()
+            }
+        }
+    }
+    
+    private func resetBoard() {
+        gamePageView.gridMaker.resetGrid()
+        shapeMaker.removeAllShapes(inView: gamePageView.gridMaker)
+        gameEngine.resetBoard()
+        configureGameSettings()
+    }
+    
+    private func userTurn(at tag: Int) {
+        let symbol = (Settings.shared.mySymbol == .cross) ? "Cross" : "Circle"
+        shapeMaker.createShape(symbol: symbol, tag: tag, inView: gamePageView.gridMaker)
+        gameEngine.player1Moves.insert(tag)
+        handleMoveResult(playerMoves: gameEngine.player1Moves, playerLabel: "Player 1")
+    }
+    
+    private func aiTurn() {
+        let aiMoveTag = gameEngine.getNextMoveForAI()
+        if let aiMoveTag = aiMoveTag {
+            let symbol = (Settings.shared.mySymbol == .cross) ? "Circle" : "Cross"
+            shapeMaker.createShape(symbol: symbol, tag: aiMoveTag, inView: gamePageView.gridMaker)
+            gameEngine.player2Moves.insert(aiMoveTag)
+            handleMoveResult(playerMoves: gameEngine.player2Moves, playerLabel: "Computer")
+        }
+    }
+    
+    private func handleMoveResult(playerMoves: Set<Int>, playerLabel: String) {
+        if let winningLine = gameEngine.checkWinningMove(playerMoves: playerMoves) {
+            let firstElement = winningLine[0]
+            let lastElement = winningLine[1]
+            DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
+                self.shapeMaker.createLineBetweenTags(fromTag: firstElement, toTag: lastElement, inView: self.gamePageView.gridMaker)
+                DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
+                    ActionAndAlertMaker.showAlert(title: "Result", subtitle: "\(playerLabel) Won")
+                    self.tableViewManager.updateRecord(with: (playerLabel))
+                    self.resetBoard()
+                }
+            }
+        } else if gameEngine.checkForDraw() {
+            DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
+                ActionAndAlertMaker.showAlert(title: "Result", subtitle: "It's a Draw")
+                self.tableViewManager.updateRecord(with: "Draw")
+                self.resetBoard()
+            }
+        } else {
+            gamePageView.userLabel.text = (playerLabel == "Player 1") ? opponentTurnLabel + " Turn" : "Player 1 Turn"
+            if Settings.shared.opponent == .ai && playerLabel == "Player 1" {
+                DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
+                    self.aiTurn()
+                }
+            }
+        }
+    }
+
+    @objc func restartGame() {
+        ActionAndAlertMaker.showAlertWithAction(title: "Restart Game", message: "Do you want to restart the game?", actionTitle: "Yes") { [weak self] in
+            self?.resetBoard()
+            self?.tableViewManager.resetRecords()
```

Comment:
If we are using optional self more than one in the bloc, then it's recommended to unwrap it first using guard let - something like below:
```suggestion
            guard let self = self else {
                return
            }

            self.resetBoard()
            self.tableViewManager.resetRecords()
```